File: conduit/tests/python/test_frame_extended_depth.py
"""Deep frame tests matching C++ test_frame_roundtrip.cpp depth.
Covers: frame_config (config field), frame_footer (checksum footer),
frame_direction (direction-qualified messages), frame_array (array payload).
"""
import pytest


# ---------------------------------------------------------------------------
# frame_footer: FooterFrame with checksum footer field
# (C++ "frame_footer: Data roundtrip with footer")
# ---------------------------------------------------------------------------

class TestFrameFooter:

    def test_data_roundtrip_with_footer(self):
        from frame_footer import Data, FooterFrame

        data_msg = Data()
        data_msg.value = 0xABCD

        frame = FooterFrame.wrap(data_msg)
        frame.checksum = 0x42

        encoded = frame.encode_bytes()

        # Wire: [msg_type:1][length:2][value:2][checksum:1] = 6 bytes
        assert len(encoded) == 6
        assert encoded[0] == 1          # msg_type
        assert encoded[5] == 0x42       # checksum (footer)

        decoded = FooterFrame.decode_bytes(encoded)
        assert decoded.msg_type == 1
        assert decoded.length == 6
        assert decoded.checksum == 0x42
        assert isinstance(decoded.payload, Data)
        assert decoded.payload.value == 0xABCD

    def test_ack_roundtrip_with_footer(self):
        from frame_footer import Ack, FooterFrame

        ack = Ack()
        ack.seq = 99

        frame = FooterFrame.wrap(ack)
        frame.checksum = 0xFF

        encoded = frame.encode_bytes()
        # Wire: [msg_type:1][length:2][seq:1][checksum:1] = 5 bytes
        assert len(encoded) == 5

        decoded = FooterFrame.decode_bytes(encoded)
        assert decoded.checksum == 0xFF
        assert isinstance(decoded.payload, Ack)
        assert decoded.payload.seq == 99

    def test_footer_double_encode(self):
        from frame_footer import Data, FooterFrame

        data_msg = Data()
        data_msg.value = 12345
        frame = FooterFrame.wrap(data_msg)
        frame.checksum = 0x42

        first = frame.encode_bytes()
        second = frame.encode_bytes()
        assert first == second


# ---------------------------------------------------------------------------
# frame_direction: DirFrame with direction-qualified messages
# (C++ "frame_direction: CommandResponse roundtrip (receive direction)")
# ---------------------------------------------------------------------------

class TestFrameDirection:

    def test_command_response_roundtrip(self):
        from frame_direction import CommandResponse, DirFrame

        msg = CommandResponse()
        msg.status = 1
        msg.detail = 500

        frame = DirFrame.wrap(msg)
        assert frame.msg_type == 1  # ID_VALUE = 1

        encoded = frame.encode_bytes()
        decoded = DirFrame.decode_bytes(encoded)

        # Decode should produce CommandResponse (receive direction)
        assert isinstance(decoded.payload, CommandResponse)
        assert decoded.payload.status == 1
        assert decoded.payload.detail == 500

    def test_heartbeat_roundtrip_bidirectional(self):
        from frame_direction import Heartbeat, DirFrame

        msg = Heartbeat()
        msg.seq = 9999

        frame = DirFrame.wrap(msg)
        assert frame.msg_type == 2

        encoded = frame.encode_bytes()
        decoded = DirFrame.decode_bytes(encoded)

        assert isinstance(decoded.payload, Heartbeat)
        assert decoded.payload.seq == 9999

    def test_message_constants_distinct(self):
        from frame_direction import CommandRequest, CommandResponse, Heartbeat

        # Same ID_VALUE but different TYPE_ID
        assert CommandRequest.ID_VALUE == 1
        assert CommandResponse.ID_VALUE == 1
        assert Heartbeat.ID_VALUE == 2
        assert CommandRequest.TYPE_ID != CommandResponse.TYPE_ID

    def test_direction_double_encode(self):
        from frame_direction import Heartbeat, DirFrame

        msg = Heartbeat()
        msg.seq = 42
        frame = DirFrame.wrap(msg)
        first = frame.encode_bytes()
        second = frame.encode_bytes()
        assert first == second


# ---------------------------------------------------------------------------
# frame_array: ArrayFrame with count="*" array payload
# (C++ "frame_array: single record roundtrip")
# ---------------------------------------------------------------------------

class TestFrameArray:

    def test_single_record_roundtrip(self):
        from frame_array import ArrayFrame, Record

        frame = ArrayFrame()
        rec = Record()
        rec.key = 1
        rec.value = 1000
        frame.payload = [rec]
        frame.msg_type = 1

        encoded = frame.encode_bytes()
        # Wire: [msg_type:1][length:2][key:1][value:2] = 6 bytes
        assert len(encoded) == 6

        decoded = ArrayFrame.decode_bytes(encoded)
        assert len(decoded.payload) == 1
        assert decoded.payload[0].key == 1
        assert decoded.payload[0].value == 1000

    def test_multiple_records_roundtrip(self):
        from frame_array import ArrayFrame, Record

        frame = ArrayFrame()
        frame.msg_type = 1
        records = []
        for i in range(5):
            rec = Record()
            rec.key = i
            rec.value = i * 100
            records.append(rec)
        frame.payload = records

        encoded = frame.encode_bytes()
        # Wire: [msg_type:1][length:2] + 5 * [key:1][value:2] = 3 + 15 = 18 bytes
        assert len(encoded) == 18

        decoded = ArrayFrame.decode_bytes(encoded)
        assert len(decoded.payload) == 5
        for i in range(5):
            assert decoded.payload[i].key == i
            assert decoded.payload[i].value == i * 100

    def test_empty_payload_roundtrip(self):
        from frame_array import ArrayFrame

        frame = ArrayFrame()
        frame.msg_type = 1
        frame.payload = []

        encoded = frame.encode_bytes()
        # Wire: [msg_type:1][length:2] = 3 bytes (header only)
        assert len(encoded) == 3

        decoded = ArrayFrame.decode_bytes(encoded)
        assert len(decoded.payload) == 0

    def test_array_double_encode(self):
        from frame_array import ArrayFrame, Record

        frame = ArrayFrame()
        frame.msg_type = 1
        rec = Record()
        rec.key = 7
        rec.value = 777
        frame.payload = [rec]

        first = frame.encode_bytes()
        second = frame.encode_bytes()
        assert first == second


# ---------------------------------------------------------------------------
# frame_config: ConfigFrame with auto="config(system-id)"
# (C++ "frame_config: config field set during encode")
# ---------------------------------------------------------------------------

class TestFrameConfig:

    def test_direct_frame_encode_decode(self):
        from frame_config import ConfigFrame, Ping

        ping = Ping()
        ping.seq = 300

        frame = ConfigFrame.wrap(ping)
        frame.system_id = 10

        encoded = frame.encode_bytes()
        # Wire: [system-id:1][msg-type:1][length:2][seq:2] = 6 bytes
        assert len(encoded) == 6
        assert encoded[0] == 10   # system-id from config
        assert encoded[1] == 1    # msg-type = Ping::ID_VALUE

        decoded = ConfigFrame.decode_bytes(encoded)
        assert decoded.system_id == 10
        assert decoded.msg_type == 1
        assert decoded.length == 6
        assert isinstance(decoded.payload, Ping)
        assert decoded.payload.seq == 300

    def test_pong_roundtrip(self):
        from frame_config import ConfigFrame, Pong

        pong = Pong()
        pong.seq = 200

        frame = ConfigFrame.wrap(pong)
        frame.system_id = 42

        encoded = frame.encode_bytes()
        assert encoded[0] == 42   # system-id
        assert encoded[1] == 2    # msg-type = Pong::ID_VALUE

        decoded = ConfigFrame.decode_bytes(encoded)
        assert decoded.system_id == 42
        assert isinstance(decoded.payload, Pong)
        assert decoded.payload.seq == 200

    def test_config_double_encode(self):
        from frame_config import ConfigFrame, Ping

        ping = Ping()
        ping.seq = 100
        frame = ConfigFrame.wrap(ping)
        frame.system_id = 7

        first = frame.encode_bytes()
        second = frame.encode_bytes()
        assert first == second

File: conduit/tests/java/TestFrameExtendedDepth.java
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Deep frame tests matching C++ test_frame_roundtrip.cpp depth.
 * Covers: frame_config (config field), frame_footer (checksum footer),
 * frame_direction (direction-qualified messages), frame_array (array payload).
 */
public class TestFrameExtendedDepth {

    // ========================================================================
    // frame_footer: FooterFrame with checksum footer field
    // (C++ "frame_footer: Data roundtrip with footer")
    // ========================================================================

    @Test
    @DisplayName("frame_footer: Data roundtrip with checksum")
    void frameFooterDataRoundtrip() {
        frame_footer.Data dataMsg = new frame_footer.Data();
        dataMsg.value = 0xABCD;

        frame_footer.FooterFrame frame = frame_footer.FooterFrame.wrap(dataMsg);
        frame.checksum = 0x42;

        byte[] encoded = frame.encodeBytes();

        // Wire: [msg_type:1][length:2][value:2][checksum:1] = 6 bytes
        assertEquals(6, encoded.length);
        assertEquals(1, encoded[0] & 0xFF);          // msg_type
        assertEquals(0x42, encoded[5] & 0xFF);        // checksum (footer)

        frame_footer.FooterFrame decoded = frame_footer.FooterFrame.decodeBytes(encoded);
        assertEquals(1, decoded.msgType);
        assertEquals(6, decoded.length);
        assertEquals(0x42, decoded.checksum);
        assertInstanceOf(frame_footer.Data.class, decoded.payload);
        assertEquals(0xABCD, ((frame_footer.Data) decoded.payload).value);
    }

    @Test
    @DisplayName("frame_footer: Ack roundtrip with checksum")
    void frameFooterAckRoundtrip() {
        frame_footer.Ack ack = new frame_footer.Ack();
        ack.seq = 99;

        frame_footer.FooterFrame frame = frame_footer.FooterFrame.wrap(ack);
        frame.checksum = (byte) 0xFF;

        byte[] encoded = frame.encodeBytes();
        // Wire: [msg_type:1][length:2][seq:1][checksum:1] = 5 bytes
        assertEquals(5, encoded.length);

        frame_footer.FooterFrame decoded = frame_footer.FooterFrame.decodeBytes(encoded);
        assertEquals(0xFF, decoded.checksum & 0xFF);
        assertInstanceOf(frame_footer.Ack.class, decoded.payload);
        assertEquals(99, ((frame_footer.Ack) decoded.payload).seq);
    }

    @Test
    @DisplayName("frame_footer: double-encode idempotency")
    void frameFooterDoubleEncode() {
        frame_footer.Data dataMsg = new frame_footer.Data();
        dataMsg.value = 12345;
        frame_footer.FooterFrame frame = frame_footer.FooterFrame.wrap(dataMsg);
        frame.checksum = 0x42;

        byte[] first = frame.encodeBytes();
        byte[] second = frame.encodeBytes();
        assertArrayEquals(first, second);
    }

    // ========================================================================
    // frame_direction: DirFrame with direction-qualified messages
    // (C++ "frame_direction: CommandResponse roundtrip")
    // ========================================================================

    @Test
    @DisplayName("frame_direction: CommandResponse roundtrip (receive)")
    void frameDirCommandResponseRoundtrip() {
        frame_direction.CommandResponse msg = new frame_direction.CommandResponse();
        msg.status = 1;
        msg.detail = 500;

        frame_direction.DirFrame frame = frame_direction.DirFrame.wrap(msg);
        assertEquals(1, frame.msgType);  // ID_VALUE = 1

        byte[] encoded = frame.encodeBytes();
        frame_direction.DirFrame decoded = frame_direction.DirFrame.decodeBytes(encoded);

        assertInstanceOf(frame_direction.CommandResponse.class, decoded.payload);
        assertEquals(1, ((frame_direction.CommandResponse) decoded.payload).status);
        assertEquals(500, ((frame_direction.CommandResponse) decoded.payload).detail);
    }

    @Test
    @DisplayName("frame_direction: Heartbeat roundtrip (bidirectional)")
    void frameDirHeartbeatRoundtrip() {
        frame_direction.Heartbeat msg = new frame_direction.Heartbeat();
        msg.seq = 9999;

        frame_direction.DirFrame frame = frame_direction.DirFrame.wrap(msg);
        assertEquals(2, frame.msgType);

        byte[] encoded = frame.encodeBytes();
        frame_direction.DirFrame decoded = frame_direction.DirFrame.decodeBytes(encoded);

        assertInstanceOf(frame_direction.Heartbeat.class, decoded.payload);
        assertEquals(9999, ((frame_direction.Heartbeat) decoded.payload).seq);
    }

    @Test
    @DisplayName("frame_direction: message constants are distinct")
    void frameDirMessageConstants() {
        assertEquals(1, frame_direction.CommandRequest.ID_VALUE);
        assertEquals(1, frame_direction.CommandResponse.ID_VALUE);
        assertEquals(2, frame_direction.Heartbeat.ID_VALUE);
        assertNotEquals(frame_direction.CommandRequest.TYPE_ID,
            frame_direction.CommandResponse.TYPE_ID);
    }

    @Test
    @DisplayName("frame_direction: double-encode idempotency")
    void frameDirDoubleEncode() {
        frame_direction.Heartbeat msg = new frame_direction.Heartbeat();
        msg.seq = 42;
        frame_direction.DirFrame frame = frame_direction.DirFrame.wrap(msg);

        byte[] first = frame.encodeBytes();
        byte[] second = frame.encodeBytes();
        assertArrayEquals(first, second);
    }

    // ========================================================================
    // frame_array: ArrayFrame with count="*" array payload
    // (C++ "frame_array: single record roundtrip")
    // ========================================================================

    @Test
    @DisplayName("frame_array: single record roundtrip")
    void frameArraySingleRecord() {
        frame_array.ArrayFrame frame = new frame_array.ArrayFrame();
        frame_array.Record rec = new frame_array.Record();
        rec.key = 1;
        rec.value = 1000;
        frame.payload.add(rec);
        frame.msgType = 1;

        byte[] encoded = frame.encodeBytes();
        // Wire: [msg_type:1][length:2][key:1][value:2] = 6 bytes
        assertEquals(6, encoded.length);

        frame_array.ArrayFrame decoded = frame_array.ArrayFrame.decodeBytes(encoded);
        assertEquals(1, decoded.payload.size());
        assertEquals(1, decoded.payload.get(0).key);
        assertEquals(1000, decoded.payload.get(0).value);
    }

    @Test
    @DisplayName("frame_array: multiple records roundtrip")
    void frameArrayMultipleRecords() {
        frame_array.ArrayFrame frame = new frame_array.ArrayFrame();
        frame.msgType = 1;
        for (int i = 0; i < 5; i++) {
            frame_array.Record rec = new frame_array.Record();
            rec.key = (byte) i;
            rec.value = i * 100;
            frame.payload.add(rec);
        }

        byte[] encoded = frame.encodeBytes();
        // Wire: [msg_type:1][length:2] + 5 * [key:1][value:2] = 3 + 15 = 18 bytes
        assertEquals(18, encoded.length);

        frame_array.ArrayFrame decoded = frame_array.ArrayFrame.decodeBytes(encoded);
        assertEquals(5, decoded.payload.size());
        for (int i = 0; i < 5; i++) {
            assertEquals(i, decoded.payload.get(i).key);
            assertEquals(i * 100, decoded.payload.get(i).value);
        }
    }

    @Test
    @DisplayName("frame_array: empty payload roundtrip")
    void frameArrayEmptyPayload() {
        frame_array.ArrayFrame frame = new frame_array.ArrayFrame();
        frame.msgType = 1;

        byte[] encoded = frame.encodeBytes();
        // Wire: [msg_type:1][length:2] = 3 bytes (header only)
        assertEquals(3, encoded.length);

        frame_array.ArrayFrame decoded = frame_array.ArrayFrame.decodeBytes(encoded);
        assertEquals(0, decoded.payload.size());
    }

    @Test
    @DisplayName("frame_array: double-encode idempotency")
    void frameArrayDoubleEncode() {
        frame_array.ArrayFrame frame = new frame_array.ArrayFrame();
        frame.msgType = 1;
        frame_array.Record rec = new frame_array.Record();
        rec.key = 7;
        rec.value = 777;
        frame.payload.add(rec);

        byte[] first = frame.encodeBytes();
        byte[] second = frame.encodeBytes();
        assertArrayEquals(first, second);
    }

    // ========================================================================
    // frame_config: ConfigFrame with auto="config(system-id)"
    // (C++ "frame_config: config field set during encode")
    // ========================================================================

    @Test
    @DisplayName("frame_config: direct frame encode/decode")
    void frameConfigDirectRoundtrip() {
        frame_config.Ping ping = new frame_config.Ping();
        ping.seq = 300;

        frame_config.ConfigFrame frame = frame_config.ConfigFrame.wrap(ping);
        frame.systemId = 10;

        byte[] encoded = frame.encodeBytes();
        // Wire: [system-id:1][msg-type:1][length:2][seq:2] = 6 bytes
        assertEquals(6, encoded.length);
        assertEquals(10, encoded[0] & 0xFF);  // system-id from config
        assertEquals(1, encoded[1] & 0xFF);   // msg-type = Ping::ID_VALUE

        frame_config.ConfigFrame decoded = frame_config.ConfigFrame.decodeBytes(encoded);
        assertEquals(10, decoded.systemId);
        assertEquals(1, decoded.msgType);
        assertEquals(6, decoded.length);
        assertInstanceOf(frame_config.Ping.class, decoded.payload);
        assertEquals(300, ((frame_config.Ping) decoded.payload).seq);
    }

    @Test
    @DisplayName("frame_config: Pong roundtrip with config")
    void frameConfigPongRoundtrip() {
        frame_config.Pong pong = new frame_config.Pong();
        pong.seq = 200;

        frame_config.ConfigFrame frame = frame_config.ConfigFrame.wrap(pong);
        frame.systemId = 42;

        byte[] encoded = frame.encodeBytes();
        assertEquals(42, encoded[0] & 0xFF);  // system-id
        assertEquals(2, encoded[1] & 0xFF);   // msg-type = Pong::ID_VALUE

        frame_config.ConfigFrame decoded = frame_config.ConfigFrame.decodeBytes(encoded);
        assertEquals(42, decoded.systemId);
        assertInstanceOf(frame_config.Pong.class, decoded.payload);
        assertEquals(200, ((frame_config.Pong) decoded.payload).seq);
    }

    @Test
    @DisplayName("frame_config: double-encode idempotency")
    void frameConfigDoubleEncode() {
        frame_config.Ping ping = new frame_config.Ping();
        ping.seq = 100;
        frame_config.ConfigFrame frame = frame_config.ConfigFrame.wrap(ping);
        frame.systemId = 7;

        byte[] first = frame.encodeBytes();
        byte[] second = frame.encodeBytes();
        assertArrayEquals(first, second);
    }
}
